paths/paths.go

```diff
@@ -59,3 +59,34 @@ func DefaultTailscaledStateFile() string {
 	}
 	return ""
 }
+
+// MkStateDir ensures that the directory dirPath exists and has the correct
+// permisions:
+// If it creates the directory, then the directory is created with the correct permissions.
+// If the directory already exists, we ensure that the permissions are set
+// correctly.
+func MkStateDir(dirPath string) error {
+	info, err := os.Stat(dirPath)
+	if err == nil && !info.IsDir() {
+		// dirPath already exists but is not a directory
+		return os.ErrInvalid
+	} else {
+		if !os.IsNotExist(err) {
+			return err
+		}
+
+		// The directory does not exist yet. Create it.
+		err = os.MkdirAll(dirPath, 0700)
+
+		// On non-Windows we're done and can just return.
+		// On Windows, when the MkdirAll succeeds, we still need to call
+		// ensureStateDirPerms to set the correct ACL.
+		if runtime.GOOS != "windows" || err != nil {
```

Comment:
I'd drop the runtime.GOOS check here and just unconditionally call ensureStateDirPerms.

If you want to avoid the `os.Chmod` call in the common case (already 0700), then pass in `info` (more commonly spelled `fi` for `FileInfo`) to `ensureStateDirPerms`.

Then this is just:

```go
      if err := os.MkdirAll(dirPath, 0700); err != nil {
            return err
      }
      return ensureStateDirPerms(dirPath, fi)
```
 
(and a lot of the comments can go away or be simplified)

And then it handles the case where due to earlier bugs or people messing around the `dirPath` was 0755.

Reply:
Since `os.MkdirAll` shouls return an error if `dirPath` is not a directory, I eliminated the `os.Stat`. So we just go ahead and do that, followed by the perm fixup.